Use a call graph for usage detection

GearsDatapacks opened the pull request:
This PR closes #2634 and fixes #3552 
Usage detection of local variables still uses the old system, because I couldn't find a way that made sense to integrate them into the call graph. Because everything else uses the call graph now though, this shouldn't cause any bugs.

lpil commented on compiler-core/src/type_/tests.rs:
```diff
@@ -213,6 +214,7 @@ macro_rules! assert_warning {
             crate::build::Target::Erlang,
             None
         );
+        assert!(!warning.is_empty());
```
Nice!!

lpil commented on compiler-core/src/reference.rs:
```diff
@@ -24,13 +25,48 @@ pub struct Reference {
+enum EntityLayer {
+    Type,
+    Value,
+    Shadowed,
```
What is this variant for? Could you document these please 🙏

lpil commented on compiler-core/src/reference.rs:
```diff
@@ -24,13 +25,48 @@ pub struct Reference {
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum EntityKind {
```
Could you document what this type is used for please 🙏

lpil commented on compiler-core/src/reference.rs:
```diff
@@ -44,51 +80,243 @@ impl ReferenceTracker {
             .push(Reference { location, kind });
     }
+
+    pub fn register_type_reference_in_call_graph(&mut self, name: EcoString) {
```
What's the difference between this and register_type_reference? It's not clear to me why one is used some times and the other other times.

lpil reviewed (commented):
Amazing!! I love this!

I've left some small notes and questions inline 🙏

GearsDatapacks commented:
I have now addressed your comments. I've also renamed a few things I forgot to rename before, such as `unused_values` -> `unused`. That's due to the old way that this code worked, so the names should be a bit more clear now.

lpil reviewed (approved):
I'm so excited for this one!! Thank you!!
